File: python/240_searchA2DMatrixII.py
class Solution:
    # @param {integer[][]} matrix
    # @param {integer} target
    # @return {boolean}
    def searchMatrix(self, matrix, target):
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return False
        n = len(matrix[0])
        for i in xrange(0, len(matrix)):
            if target < matrix[i][0]:
                return False
            if target > matrix[i][n-1]:
                continue
            left = 0
            right = n - 1
            while(left <= right):
                mid = (left+right)/2
                if matrix[i][mid] > target:
                    right = mid - 1
                elif matrix[i][mid] < target:
                    left = mid + 1
                else:
                    return True
        return False

File: cpp_solutions/240_searchA2DMatrixII.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.size() == 0 || matrix[0].size() == 0){
            return false;
        }
        
        int m = matrix.size();
        int n = matrix[0].size();
        //int i = 0;
        
        for (int i = 0; i < m; i++){
            if (target < matrix[i][0]){
                return false;
            }
            
            if (target > matrix[i][n-1]){
                continue;
            }
            
            int left = 0;
            int right = n-1;
            
            while(left <= right){
                int middle = (left+right) >> 1; // (left + right)/2
                if (matrix[i][middle]> target){
                    right = middle - 1;
                    //continue;
                }
                else if(matrix[i][middle] < target){
                    left = middle + 1;
                }
                else{
                    return true;
                }
            }
        }
        return false;
    }
};